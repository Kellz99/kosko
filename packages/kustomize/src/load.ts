import { LoadOptions, loadString, Manifest } from "@kosko/yaml";
import {
  spawn,
  booleanArg,
  stringArg,
  stringArrayArg
} from "@kosko/exec-utils";

const BUILD_COMMANDS = [
  ["kustomize", "build"],
  ["kubectl", "kustomize"]
];

export interface KustomizeOptions extends LoadOptions {
  /**
   * The path to a directory containing `kustomization.yaml`, or a Git repository
   * URL with a suffix specifying same with respect to the repository root.
   */
  path: string;

  /**
   * Use the uid and gid of the command executor to run the function in the container.
   */
  asCurrentUser?: boolean;

  /**
   * Enable kustomize plugins.
   */
  enableAlphaPlugins?: boolean;

  /**
   * Enable support for exec functions (raw executables); do not use for untrusted configs! (Alpha)
   */
  enableExec?: boolean;

  /**
   * Enable use of the Helm chart inflator generator.
   */
  enableHelm?: boolean;

  /**
   * Enable adding `app.kubernetes.io/managed-by` label.
   */
  enableManagedByLabel?: boolean;

  /**
   * Enable support for starlark functions. (Alpha)
   */
  enableStar?: boolean;

  /**
   * A list of environment variables to be used by functions.
   */
  env?: Record<string, string>;

  /**
   * Helm command (path to executable). (default `helm`)
   */
  helmCommand?: string;

  /**
   * If set to `LoadRestrictionsNone`, local kustomizations may load files from
   * outside their root. This does, however, break the relocatability of the
   * kustomization. (default `LoadRestrictionsRootOnly`)
   */
  loadRestrictor?: string;

  /**
   * A list of storage options read from the filesystem.
   */
  mount?: string[];

  /**
   * Enable network access for functions that declare it.
   */
  network?: boolean;

  /**
   * The docker network to run the container in. (default `bridge`)
   */
  networkName?: string;

  /**
   * Reorder the resources just before output. Use `legacy` to apply a legacy
   * reordering (Namespaces first, Webhooks last, etc). Use `none` to suppress a
   * final reordering. (default `legacy`)
   */
  reorder?: string;

  /**
   * The command to build Kustomize files. By default, it uses `kustomize build`
   * when the `kustomize` executable exists, otherwise fallback to `kubectl kustomize`.
   */
  buildCommand?: readonly string[];
}

export function loadKustomize({
  path,
  asCurrentUser,
  enableAlphaPlugins,
  enableExec,
  enableHelm,
  enableManagedByLabel,
  enableStar,
  env,
  helmCommand,
  loadRestrictor,
  mount,
  network,
  networkName,
  reorder,
  transform,
  buildCommand
}: KustomizeOptions) {
  async function build(cmd: readonly string[]): Promise<Manifest[]> {
    const [command, ...args]: string[] = [
      ...cmd,
      path,
      ...booleanArg("as-current-user", asCurrentUser),
      ...booleanArg("enable-alpha-plugins", enableAlphaPlugins),
      ...booleanArg("enable-exec", enableExec),
      ...booleanArg("enable-helm", enableHelm),
      ...booleanArg("enable-managedby-label", enableManagedByLabel),
      ...booleanArg("enable-star", enableStar),
      ...(env
        ? stringArrayArg(
            "env",
            Object.entries(env).map(([k, v]) => `${k}=${v}`)
          )
        : []),
      ...stringArg("helm-command", helmCommand),
      ...stringArg("load-restrictor", loadRestrictor),
      ...stringArrayArg("mount", mount),
      ...booleanArg("network", network),
      ...stringArg("network-name", networkName),
      ...stringArg("reorder", reorder)
    ];

    const { stdout } = await spawn(command, args);

    return loadString(stdout, { transform });
  }

  return async () => {
    if (buildCommand) {
      return build(buildCommand);
    }

    for (const cmd of BUILD_COMMANDS) {
      try {
        return await build(cmd);
      } catch (err: any) {
        if (err.code !== "ENOENT") {
          throw err;
        }
      }
    }

    throw new Error(
      `"loadKustomize" requires either kustomize or kubectl CLI installed in your environment. More info: https://kosko.dev/docs/components/loading-kustomize`
    );
  };
}
