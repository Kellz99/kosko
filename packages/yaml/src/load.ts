import { safeLoadAll } from "js-yaml";
import fs from "fs";
import { promisify } from "util";
import fetch, { RequestInfo, RequestInit } from "node-fetch";
import { getResourceModule, ResourceKind } from "./module";
import Debug from "./debug";

const debug = Debug.extend("load");
const readFile = promisify(fs.readFile);

export interface Manifest extends ResourceKind {
  [key: string]: any;
}

type ManifestConstructor = new (data: Manifest) => Manifest;

export interface LoadOptions {
  transform?(manifest: Manifest): Manifest | null | undefined;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && !Array.isArray(value);
}

function isManifest(value: Record<string, unknown>): value is Manifest {
  return (
    typeof value.apiVersion === "string" &&
    !!value.apiVersion &&
    typeof value.kind === "string" &&
    !!value.kind
  );
}

function getConstructor(res: ResourceKind): ManifestConstructor | undefined {
  const mod = getResourceModule(res);

  if (!mod) {
    debug("No resource modules for", res);
    return;
  }

  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return require(mod.path)[mod.export];
  } catch {
    debug("Failed to import the resource module", mod);
    return;
  }
}

/**
 * Load a Kubernetes YAML file from a string.
 *
 * ## Examples
 *
 * ### Transform manifests
 *
 * ```ts
 * loadString('', {
 *   transform(manifest) {
 *     if (manifest.apiVersion === "apps/v1" && manifest.kind === "Deployment") {
 *       manifest.spec.replicas = 3;
 *     }
 *
 *     return manifest;
 *   }
 * });
 * ```
 *
 * ### Filter manifests
 *
 * Values are removed from array if `transform` function returns a falsy value,
 * such as `null` or `undefined`.
 *
 * ```ts
 * loadString('', {
 *   transform(manifest) {
 *     if (manifest.metadata.name === 'foo') {
 *       return null;
 *     }
 *
 *     return manifest;
 *   }
 * });
 * ```
 */
export function loadString(
  content: string,
  { transform = (x) => x }: LoadOptions = {}
): ReadonlyArray<Manifest> {
  const input = safeLoadAll(content).filter((x) => x != null);
  const manifests: Manifest[] = [];

  for (const entry of input) {
    if (!isRecord(entry)) {
      throw new Error(`The value must be an object: ${JSON.stringify(entry)}`);
    }

    if (!isManifest(entry)) {
      throw new Error(
        `apiVersion and kind are required: ${JSON.stringify(entry)}`
      );
    }

    const Constructor = getConstructor(entry);
    const manifest = transform(Constructor ? new Constructor(entry) : entry);

    if (manifest) {
      manifests.push(manifest);
    }
  }

  return manifests;
}

/**
 * Load a Kubernetes YAML file from path.
 *
 * @param path Path to the Kubernetes YAML file.
 * @param options
 */
export function loadFile(path: string, options?: LoadOptions) {
  return async (): Promise<ReadonlyArray<Manifest>> => {
    const content = await readFile(path, "utf-8");
    debug("File loaded from: %s", path);

    return loadString(content, options);
  };
}

/**
 * Load a Kubernetes YAML file from url.
 *
 * @param url URL to the Kubernetes YAML file.
 * @param options [Options](https://github.com/node-fetch/node-fetch#options) for the HTTP(S) request.
 */
export function loadUrl(
  url: RequestInfo,
  options: LoadOptions & RequestInit = {}
): () => Promise<ReadonlyArray<Manifest>> {
  const { transform, ...init } = options;

  return async (): Promise<ReadonlyArray<Manifest>> => {
    const res = await fetch(url, init);
    debug(`Fetch "%s": status=%d`, url, res.status);

    if (!res.ok) {
      throw new Error(`Failed to fetch YAML file from: ${url}`);
    }

    return loadString(await res.text(), { transform });
  };
}
