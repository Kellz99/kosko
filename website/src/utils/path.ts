export function isAbsolute(path: string): boolean {
  return path.startsWith("/");
}

export function isRelative(path: string): boolean {
  return path.startsWith("./") || path.startsWith("../");
}

export function dirname(path: string): string {
  const index = path.lastIndexOf("/");
  if (index === -1) return path;
  return path.substring(0, index - 1) || "/";
}

export function extname(path: string): string {
  const index = path.lastIndexOf(".");
  if (index === -1) return "";
  return path.substring(index);
}

export function basename(path: string, ext?: string): string {
  const index = path.lastIndexOf("/");
  const base = index === -1 ? path : path.substring(index + 1);

  if (base.endsWith(ext)) {
    return base.substring(0, base.length - ext.length);
  }

  return base;
}

export function relative(from: string, to: string): string {
  const fromParts = from.split("/").filter(Boolean);
  const toParts = to.split("/").filter(Boolean);

  while (fromParts[0] && toParts[0]) {
    fromParts.shift();
    toParts.shift();
  }

  while (toParts[0] === "." || toParts[0] === "..") {
    const toPart = toParts.shift();

    if (toPart === "..") {
      fromParts.pop();
    }
  }

  while (fromParts.pop()) {
    toParts.unshift("..");
  }

  return toParts.join("/");
}
