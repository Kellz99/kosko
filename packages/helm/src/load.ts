import { LoadOptions, loadString, Manifest } from "@kosko/yaml";
import tmp from "tmp-promise";
import { writeFile, readFile, mkdir } from "node:fs/promises";
import {
  spawn,
  booleanArg,
  stringArg,
  stringArrayArg
} from "@kosko/exec-utils";
import stringify from "fast-safe-stringify";
import { getErrorCode, isRecord } from "@kosko/common-utils";
import getCacheDir from "cachedir";
import { createHash } from "node:crypto";
import { dirname, join } from "node:path";
import { env } from "node:process";
import yaml from "js-yaml";
import { fileExists, move } from "./fs";

const FILE_EXIST_ERROR_CODES = new Set(["EEXIST", "ENOTEMPTY"]);
const OCI_PREFIX = "oci://";

const defaultCacheDir = getCacheDir("kosko-helm");

/**
 * @public
 */
export interface CacheOptions {
  /**
   * When cache is enabled, the chart is pulled and stored in the cache
   * directory. Although Helm has its own cache, implementing our own cache
   * is faster. Local charts are never cached.
   *
   * @defaultValue `true`
   */
  enabled?: boolean;

  /**
   * The path of the cache directory. You can also use `KOSKO_HELM_CACHE_DIR`
   * environment variable to set the cache directory. This option always takes
   * precedence over the environment variable.
   *
   * @defaultValue
   * - Linux: `$XDG_CACHE_HOME/kosko-helm` or `~/.cache/kosko-helm`
   * - macOS: `~/Library/Caches/kosko-helm`
   * - Windows: `$LOCALAPPDATA/kosko-helm` or `~/AppData/Local/kosko-helm`
   */
  dir?: string;
}

/**
 * @public
 */
export interface PullOptions {
  /**
   * The path of a local chart or the name of a remote chart.
   */
  chart: string;

  /**
   * Verify certificates of HTTPS-enabled servers using this CA bundle.
   */
  caFile?: string;

  /**
   * Identify HTTPS client using this SSL certificate file.
   */
  certFile?: string;

  /**
   * Use development versions, too. Equivalent to version '\>0.0.0-0'. If
   * `version` is set, this is ignored.
   */
  devel?: boolean;

  /**
   * Skip tls certificate checks for the chart download.
   */
  insecureSkipTlsVerify?: boolean;

  /**
   * Identify HTTPS client using this SSL key file.
   */
  keyFile?: string;

  /**
   * Location of public keys used for verification.
   *
   * @defaultValue `~/.gnupg/pubring.gpg`
   */
  keyring?: string;

  /**
   * Pass credentials to all domains.
   */
  passCredentials?: boolean;

  /**
   * Chart repository password where to locate the requested chart.
   */
  password?: string;

  /**
   * Chart repository url where to locate the requested chart.
   */
  repo?: string;

  /**
   * Chart repository username where to locate the requested chart.
   */
  username?: string;

  /**
   * Verify the package before using it.
   */
  verify?: boolean;

  /**
   * Specify the exact chart version to use. If this is not specified, the
   * latest version is used.
   */
  version?: string;

  /**
   * Cache options.
   */
  cache?: CacheOptions;
}

/**
 * @public
 */
export interface TemplateOptions {
  /**
   * Name of the release.
   */
  name?: string;

  /**
   * Kubernetes API versions used for `Capabilities.APIVersions`.
   */
  apiVersions?: string[];

  /**
   * Run helm dependency update before installing the chart.
   */
  dependencyUpdate?: boolean;

  /**
   * Add a custom description.
   */
  description?: string;

  /**
   * Generate the name (and omit the `name` parameter).
   */
  generateName?: boolean;

  /**
   * Include CRDs in the templated output.
   */
  includeCrds?: boolean;

  /**
   * Set `.Release.IsUpgrade` instead of `.Release.IsInstall`.
   */
  isUpgrade?: boolean;

  /**
   * Kubernetes version used for `Capabilities.KubeVersion`.
   */
  kubeVersion?: string;

  /**
   * Specify template used to name the release.
   */
  nameTemplate?: string;

  /**
   * Namespace scope for this request.
   */
  namespace?: string;

  /**
   * Prevent hooks from running during install.
   */
  noHooks?: boolean;

  /**
   * The path to an executable to be used for post rendering. If it exists in
   * `$PATH`, the binary will be used, otherwise it will try to look for the
   * executable at the given path.
   */
  postRenderer?: string;

  /**
   * Arguments to the post-renderer.
   *
   * @defaultValue `[]`
   */
  postRendererArgs?: string[];

  /**
   * Skip tests from templated output.
   */
  skipTests?: boolean;

  /**
   * Time to wait for any individual Kubernetes operation (like Jobs for hooks)
   *
   * @defaultValue `5m0s`
   */
  timeout?: string;

  /**
   * Specify values.
   */
  values?: unknown;
}

function removeBase64Padding(str: string): string {
  const index = str.indexOf("=");
  return index === -1 ? str : str.substring(0, index);
}

function genHash(value: string): string {
  const hash = createHash("sha1");

  hash.write(value);

  return removeBase64Padding(hash.digest("base64url"));
}

function genObjectHash(value: unknown): string {
  return genHash(stringify(value));
}

async function runHelm(args: readonly string[]) {
  try {
    return await spawn("helm", args);
  } catch (err) {
    if (getErrorCode(err) !== "ENOENT") throw err;

    throw new Error(
      `"loadChart" requires Helm CLI installed in your environment. More info: https://kosko.dev/docs/components/loading-helm-chart`
    );
  }
}

function getChartManifestPath(path: string) {
  return join(path, "Chart.yaml");
}

function chartManifestExists(path: string) {
  return fileExists(getChartManifestPath(path));
}

async function isLocalChart(
  options: Pick<PullOptions, "repo" | "chart">
): Promise<boolean> {
  // If repo is set, it's a remote chart
  if (options.repo) return false;

  // OCI charts are always remote
  if (options.chart.startsWith(OCI_PREFIX)) return false;

  return chartManifestExists(options.chart);
}

function getChartBaseName(chart: string): string {
  const index = chart.lastIndexOf("/");

  return index === -1 ? chart : chart.substring(index + 1);
}

async function getChartMetadata(
  chart: string
): Promise<Record<string, unknown> | undefined> {
  try {
    const content = await readFile(getChartManifestPath(chart), "utf8");
    const metadata = yaml.load(content);

    if (isRecord(metadata)) return metadata;
  } catch (err) {
    if (getErrorCode(err) !== "ENOENT") throw err;
  }
}

function getPullArgs(options: PullOptions): string[] {
  return [
    options.chart,
    ...stringArg("ca-file", options.caFile),
    ...stringArg("cert-file", options.certFile),
    ...booleanArg("devel", options.devel),
    ...booleanArg("insecure-skip-tls-verify", options.insecureSkipTlsVerify),
    ...stringArg("key-file", options.keyFile),
    ...stringArg("keyring", options.keyring),
    ...booleanArg("pass-credentials", options.passCredentials),
    ...stringArg("password", options.password),
    ...stringArg("repo", options.repo),
    ...stringArg("username", options.username),
    ...booleanArg("verify", options.verify),
    ...stringArg("version", options.version)
  ];
}

async function moveChartToCacheDir(src: string, dest: string): Promise<void> {
  // Skip if the chart already exists in the cache directory
  if (await chartManifestExists(dest)) return;

  await mkdir(dirname(dest), { recursive: true });

  try {
    await move(src, dest);
  } catch (err) {
    const code = getErrorCode(err);

    if (!code) throw err;

    // If the target already exists, it probably means that another
    // process has already pulled the chart. In this case, we can ignore the
    // error and return the cache path.
    if (FILE_EXIST_ERROR_CODES.has(code)) {
      return;
    }

    // Windows throws EPERM error when the target already exists. In this case,
    // we will try to check if the `Chart.yaml` exists in the target directory.
    //
    // https://github.com/nodejs/node/issues/29481
    if (code === "EPERM" && (await chartManifestExists(dest))) {
      return;
    }

    throw err;
  }
}

async function pullChart(options: PullOptions): Promise<string | undefined> {
  // Skip cache if disabled
  if (options.cache?.enabled === false) return;

  // Skip cache if version is not set
  if (!options.version) return;

  // Skip cache if it's a local chart
  if (await isLocalChart(options)) return;

  const cacheDir =
    options.cache?.dir || env.KOSKO_HELM_CACHE_DIR || defaultCacheDir;
  const indexHash = genObjectHash({
    c: options.chart,
    d: options.devel,
    r: options.repo,
    v: options.version
  });
  const indexPath = join(cacheDir, "index-" + indexHash);

  // Read the index file to get the cache path
  try {
    const content = await readFile(indexPath, "utf8");
    if (content) {
      // The content read by current process might be empty due to race condition.
      return join(cacheDir, content);
    }
  } catch (err) {
    if (getErrorCode(err) !== "ENOENT") throw err;
  }

  // Create a temporary directory for the chart, because when there are multiple
  // processes pulling the same chart, sometimes Helm fails because files already
  // exist in the cache directory.
  const tmpDir = await tmp.dir({ prefix: "kosko-helm", unsafeCleanup: true });

  try {
    // Pull the chart
    await runHelm([
      "pull",
      ...getPullArgs(options),
      "--untar",
      "--untardir",
      tmpDir.path
    ]);

    const chartDir = join(tmpDir.path, getChartBaseName(options.chart));

    // Get chart version
    const chartMeta = await getChartMetadata(chartDir);
    const chartVersion = chartMeta?.version;

    if (typeof chartVersion !== "string") return;

    const chartHash = genObjectHash({
      c: options.chart,
      r: options.repo,
      v: chartVersion
    });
    const dest = join(cacheDir, chartHash);

    // Move the chart to the cache directory
    await moveChartToCacheDir(chartDir, dest);

    // Write index file
    await writeFile(indexPath, chartHash);

    return dest;
  } finally {
    await tmpDir.cleanup();
  }
}

async function writeValues(values: unknown) {
  const file = await tmp.file();

  await writeFile(file.path, stringify(values));

  return file;
}

async function renderChart(options: ChartOptions) {
  const args: string[] = [
    "template",
    ...(options.name ? [options.name] : []),
    ...getPullArgs(options),
    ...stringArrayArg("api-versions", options.apiVersions),
    ...booleanArg("dependency-update", options.dependencyUpdate),
    ...stringArg("description", options.description),
    ...booleanArg("generate-name", options.generateName),
    ...booleanArg("include-crds", options.includeCrds),
    ...booleanArg("is-upgrade", options.isUpgrade),
    ...stringArg("kube-version", options.kubeVersion),
    ...stringArg("name-template", options.nameTemplate),
    ...stringArg("namespace", options.namespace),
    ...booleanArg("no-hooks", options.noHooks),
    ...stringArg("post-renderer", options.postRenderer),
    ...stringArrayArg("post-renderer-args", options.postRendererArgs),
    ...booleanArg("skip-tests", options.skipTests),
    ...stringArg("timeout", options.timeout)
  ];
  let valueFile: tmp.FileResult | undefined;

  if (options.values) {
    valueFile = await writeValues(options.values);
    args.push("--values", valueFile.path);
  }

  try {
    return await runHelm(args);
  } finally {
    await valueFile?.cleanup();
  }
}

/**
 * @public
 */
export interface ChartOptions
  extends LoadOptions,
    PullOptions,
    TemplateOptions {}

/**
 * @public
 */
export function loadChart(options: ChartOptions): () => Promise<Manifest[]> {
  const { transform, ...opts } = options;

  return async () => {
    const cachedChart = await pullChart(opts);
    const { stdout } = await renderChart({
      ...opts,
      ...(cachedChart && { chart: cachedChart, repo: undefined })
    });

    return loadString(stdout, { transform });
  };
}
