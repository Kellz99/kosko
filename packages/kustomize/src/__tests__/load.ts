import { join } from "path";
import { loadKustomize } from "../load";

const FIXTURE_DIR = join(__dirname, "../__fixtures__");

jest.setTimeout(15000);

test("local path", async () => {
  const result = loadKustomize({ path: join(FIXTURE_DIR, "hello-world") });
  await expect(result()).resolves.toMatchSnapshot();
});

test("remote url", async () => {
  const result = loadKustomize({
    path: "github.com/kustless/kustomize-examples"
  });

  await expect(result()).resolves.toMatchSnapshot();
});

test("helm chart", async () => {
  const result = loadKustomize({
    path: join(FIXTURE_DIR, "helm"),
    enableHelm: true
  });

  await expect(result()).resolves.toMatchSnapshot();
});
