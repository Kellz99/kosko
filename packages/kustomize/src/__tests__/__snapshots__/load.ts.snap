// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helm chart 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "data": Object {
      "server-blocks-paths.conf": "include  \\"/opt/bitnami/nginx/conf/server_blocks/ldap/*.conf\\";
include  \\"/opt/bitnami/nginx/conf/server_blocks/common/*.conf\\";",
    },
    "kind": "ConfigMap",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "foobar",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "nginx",
        "helm.sh/chart": "nginx-9.5.16",
      },
      "name": "foobar-nginx-server-block",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "foobar",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "nginx",
        "helm.sh/chart": "nginx-9.5.16",
      },
      "name": "foobar-nginx",
    },
    "spec": Object {
      "externalTrafficPolicy": "Cluster",
      "ports": Array [
        Object {
          "name": "http",
          "port": 80,
          "targetPort": "http",
        },
      ],
      "selector": Object {
        "app.kubernetes.io/instance": "foobar",
        "app.kubernetes.io/name": "nginx",
      },
      "type": "LoadBalancer",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "foobar",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "nginx",
        "helm.sh/chart": "nginx-9.5.16",
      },
      "name": "foobar-nginx",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "app.kubernetes.io/instance": "foobar",
          "app.kubernetes.io/name": "nginx",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app.kubernetes.io/instance": "foobar",
            "app.kubernetes.io/managed-by": "Helm",
            "app.kubernetes.io/name": "nginx",
            "helm.sh/chart": "nginx-9.5.16",
          },
        },
        "spec": Object {
          "affinity": Object {
            "nodeAffinity": null,
            "podAffinity": null,
            "podAntiAffinity": Object {
              "preferredDuringSchedulingIgnoredDuringExecution": Array [
                Object {
                  "podAffinityTerm": Object {
                    "labelSelector": Object {
                      "matchLabels": Object {
                        "app.kubernetes.io/instance": "foobar",
                        "app.kubernetes.io/name": "nginx",
                      },
                    },
                    "namespaces": Array [
                      "default",
                    ],
                    "topologyKey": "kubernetes.io/hostname",
                  },
                  "weight": 1,
                },
              ],
            },
          },
          "automountServiceAccountToken": false,
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "BITNAMI_DEBUG",
                  "value": "false",
                },
              ],
              "image": "docker.io/bitnami/nginx:1.21.4-debian-10-r27",
              "imagePullPolicy": "IfNotPresent",
              "livenessProbe": Object {
                "failureThreshold": 6,
                "periodSeconds": 10,
                "successThreshold": 1,
                "tcpSocket": Object {
                  "port": "http",
                },
                "timeoutSeconds": 5,
              },
              "name": "nginx",
              "ports": Array [
                Object {
                  "containerPort": 8080,
                  "name": "http",
                },
              ],
              "readinessProbe": Object {
                "failureThreshold": 3,
                "initialDelaySeconds": 5,
                "periodSeconds": 5,
                "successThreshold": 1,
                "tcpSocket": Object {
                  "port": "http",
                },
                "timeoutSeconds": 3,
              },
              "resources": Object {
                "limits": Object {},
                "requests": Object {},
              },
              "volumeMounts": null,
            },
          ],
          "serviceAccountName": "default",
          "shareProcessNamespace": false,
          "volumes": Array [
            Object {
              "configMap": Object {
                "items": Array [
                  Object {
                    "key": "server-blocks-paths.conf",
                    "path": "server-blocks-paths.conf",
                  },
                ],
                "name": "foobar-nginx-server-block",
              },
              "name": "nginx-server-block-paths",
            },
          ],
        },
      },
    },
  },
]
`;

exports[`local path 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "data": Object {
      "altGreeting": "Good Morning!",
      "enableRisky": "false",
    },
    "kind": "ConfigMap",
    "metadata": Object {
      "labels": Object {
        "app": "hello",
      },
      "name": "the-map",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "app": "hello",
      },
      "name": "the-service",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 8666,
          "protocol": "TCP",
          "targetPort": 8080,
        },
      ],
      "selector": Object {
        "app": "hello",
        "deployment": "hello",
      },
      "type": "LoadBalancer",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "app": "hello",
      },
      "name": "the-deployment",
    },
    "spec": Object {
      "replicas": 3,
      "selector": Object {
        "matchLabels": Object {
          "app": "hello",
          "deployment": "hello",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app": "hello",
            "deployment": "hello",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "command": Array [
                "/hello",
                "--port=8080",
                "--enableRiskyFeature=$(ENABLE_RISKY)",
              ],
              "env": Array [
                Object {
                  "name": "ALT_GREETING",
                  "valueFrom": Object {
                    "configMapKeyRef": Object {
                      "key": "altGreeting",
                      "name": "the-map",
                    },
                  },
                },
                Object {
                  "name": "ENABLE_RISKY",
                  "valueFrom": Object {
                    "configMapKeyRef": Object {
                      "key": "enableRisky",
                      "name": "the-map",
                    },
                  },
                },
              ],
              "image": "monopole/hello:1",
              "name": "the-container",
              "ports": Array [
                Object {
                  "containerPort": 8080,
                },
              ],
            },
          ],
        },
      },
    },
  },
]
`;

exports[`remote url 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "data": Object {
      "altGreeting": "Good Morning!",
      "enableRisky": "false",
    },
    "kind": "ConfigMap",
    "metadata": Object {
      "name": "remote-cm",
    },
  },
]
`;
